appcues/src/main/java/com/appcues/debugger/screencapture/ScreenCaptureProcessor.kt

```diff
@@ -48,14 +60,87 @@ internal class ScreenCaptureProcessor(
         }
     }
 
-    @Suppress("UnusedPrivateMember")
-    suspend fun save(capture: Capture, token: String): ResultOf<Capture, Error> {
+    suspend fun save(capture: Capture, token: String) =
+        // Saving a screen is a 4-step chain. This is implemented here as a sequence of calls, chaining
+        // a success continuation on each call to move to the next step. If any step fails, the RemoteError
+        // is bubbled up to the caller of this function instead of the successful result, and an error
+        // can be shown and handled.
+
+        // step 1 - get the settings, with the path to customer API
+        getSettings { sdkSettings ->
```

Comment:
wondering if you are interested in changing this piece of code to look like this:


```kotlin

    suspend fun save(capture: Capture, token: String): ResultOf<Capture, RemoteError> {
        // Saving a screen is a 4-step chain. This is implemented here as a sequence of calls, chaining
        // a success continuation on each call to move to the next step. If any step fails, the RemoteError
        // is bubbled up to the caller of this function instead of the successful result, and an error
        // can be shown and handled.
        return try {
            // step 1 - get the settings, with the path to customer API
            sdkSettingsRemoteSource.getSettings()
                // step 2 - use the customer API path to get the link to upload the screenshot
                .getUploadPath(capture, token)
                // step 3 - upload the screenshot
                .uploadImage(capture)
                // step 4 - update the screenshot link and save the screen
                .saveScreen(capture, token)
        } catch (e: ScreenCaptureSaveException) {
            Failure(e.error)
        }
    }
```

Reply:
I achieved this by removing the callbacks on success and chaining the return to one that made sense for the next step, on every remote call I would check if the result is failure and in case it is I throw this private ScreenCaptureSaveException that we try/catch on save to ensure that any exception here is captured and returned as failure later.

```kotlin
    private class ScreenCaptureSaveException(val error: RemoteError) : Exception("ScreenCaptureException: $error")

    // step 1 - get the settings, with the path to customer API
    private suspend fun SdkSettingsRemoteSource.getSettings(): SdkSettingsResponse {
        return sdkSettings().let {
            when (it) {
                is Failure -> throw ScreenCaptureSaveException(it.reason)
                is Success -> it.value
            }
        }
    }
```

Reply:
Thanks for the idea! I do think it reads more cleanly this way, giving it a shot.